tools/2025-01-gitops-technical-design.md

```diff
@@ -0,0 +1,435 @@
+
+#### Creating the `DeploymentTemplate` Resource
+
+The controller will create (or update) a `DeploymentTemplate` resource on the cluster with the compiled Bicep file as the template:
+```go
+  // Create the DeploymentTemplate resource
+  deploymentTemplate := &radappiov1alpha3.DeploymentTemplate{
+    ObjectMeta: metav1.ObjectMeta{
+      Name:      radiusConfig.Name,
+      Namespace: radiusConfig.Namespace,
+    },
+    Spec: radappiov1alpha3.DeploymentTemplateSpec{
+      Template: string(compiledBicep),
+    },
+  }
+
+  // Create or update the DeploymentTemplate resource
+  if err := r.CreateOrUpdate(ctx, deploymentTemplate); err != nil {
+    logger.Error(err, "unable to create or update DeploymentTemplate resource")
+    return ctrl.Result{}, err
+  }
+```
+
+#### Deleting Stale `DeploymentTemplate` Resources
+
+The controller will list all `DeploymentTemplate` resources in the cluster that have been created from this Git repository and delete any resources that are not present in the Git repository:
+```go
+	// List all DeploymentTemplates on the cluster that are from the same git repository
+	deploymentTemplates := &radappiov1alpha3.DeploymentTemplateList{}
+	err = r.Client.List(ctx, deploymentTemplates, client.MatchingFields{deploymentTemplateRepositoryField: repository.Name})
+	if err != nil {
+		logger.Error(err, "unable to list deployment templates")
+		return ctrl.Result{}, err
+	}
+
+	// For all of the DeploymentTemplates on the cluster, check if the bicep file
+	// that it was created from is still present in the git repository. If not, delete the
+	// DeploymentTemplate from the cluster.
+	for _, deploymentTemplate := range deploymentTemplates.Items {
+		if _, err := r.FileSystem.Stat(path.Join(tmpDir, deploymentTemplate.Name)); os.IsNotExist(err) {
+			// File does not exist in the git repository,
+			// delete the DeploymentTemplate from the cluster
+			logger.Info("Deleting DeploymentTemplate", "name", deploymentTemplate.Name)
+			if err := r.Client.Delete(ctx, &deploymentTemplate); err != nil {
+				logger.Error(err, "unable to delete deployment template")
+				return ctrl.Result{}, err
+			}
+
+			logger.Info("Deleted DeploymentTemplate", "name", deploymentTemplate.Name)
+		} else if err != nil {
+			logger.Error(err, "failed to check if file exists")
+			return ctrl.Result{}, err
+		}
+	}
+```
+
+### Detailed Design: Bicep Compilation
+
+The Bicep compilation will be done using the Bicep CLI. The Bicep CLI itself will be present in a container image that is mounted as a volume in the controller pod. The controller will use the Bicep CLI to compile the Bicep file and return the compiled template as a string.
+
+`deploy/images/bicep/Dockerfile`:
+```Dockerfile
+FROM alpine:latest
+
+ARG TARGETARCH
+
+RUN apk --no-cache add curl
+
+WORKDIR /
+
+RUN curl -Lo bicep https://github.com/Azure/bicep/releases/latest/download/bicep-linux-x64 \
```

Comment:
Have we thought about supporting different architectures? This is scoped to only Linux images

Reply:
yes, the design should support all architectures. this is a typo

Reply:
Or cases where the user may want to bring their own specific version of Bicep?